### PR TITLE
Initial implementation of stateless RNG APIs

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #178183
* #177898
* #177689
* #178160
* #177882
* #177692
* #177230
* __->__ #177229

Initial implementation for a JAX-like stateless RNG API, starting with:
* `torch.func._random.key(seed)`
* `torch.func._random.split(key, num)`
* `torch.func._random.fold_in(key, data)`

Both `split()` and `fold_in()` support arbitrarily-batched keys. The former will add a new left-most batch dimension while the latter maintains the input shape. 

In the details, both `split()` and `fold_in()` sample 128-bits of randomness at each (seed, offset) pair to produce a new (seed, offset) pair. 64-bits of the sampled randomness are used for the seed and 64-bits are used for the offset.

Underneath, the `split()` and `fold_in()` kernels utilize a custom stateless Philox-4x32-10 round implementation. This implementation fixes `subsequence=0` permanently to ensure consistent random number generation across # of CUDA threads and thus across GPU devices, input shapes, etc.

Example API usage:
```python
import torch.func._random as r

# === Randomness using split() ===
key = r.key(42, device="cuda")
for i in range(3):
  new_key, subkey = r.split(key)
  del key  # The old key is consumed by split() -- we must never use it again.

  val = r.normal(key, 3)
  del subkey  # The subkey is consumed by normal().

  print(f"draw {i}: {val}")
  key = new_key  # new_key is safe to use in the next iteration.

# === Less involved and mostly equivalent to the above ===
key = r.key(42, device="cuda")
for i in range(3):
  key, subkey = r.split(key)
  val = r.normal(subkey, 3)
  print(f"draw {i}: {val}")

# === Alternative approach using split with explicit num ===
key = r.key(42, device="cuda")
keys = r.split(key, num=4)
key, subkeys = keys[0], keys[1:]
vals = [r.normal(subkey, 3) for subkey in subkeys]
print(f"draw: {vals}")

# === Alternative approach using fold_in() ===
key = r.key(42, device="cuda")
for i in range(3):
    val = r.normal(r.fold_in(key, i), 3)
    print(f"draw {i}: {val}")
# This is needed for future randomness calls to avoid key reuse
key = r.fold_in(key, 3)

# === Batch invariance example (per-sample keys) ===
key = r.key(42, device="cuda")
B = 64
# 1 key for future randomness and the other B for per-batch-item usage.
# key shape: (2,)
# batch_key shape: (B, 1, 2)
keys = r.split(key, num=B+1)
key, batch_key = keys[0], keys[1:].unsqueeze(-2)
# each set of 32 is generated via a corresponding batch item key
x = r.normal(batch_key, B, 32)
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/PhiloxKeySplit.cu [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    for (int64_t split_idx = split_start; split_idx < split_end; split_idx++) {
+      uint32_t r0 = curand(&state);
+      uint32_t r1 = curand(&state);
+      uint32_t r2 = curand(&state);
+      uint32_t r3 = curand(&state);
+
+      uint64_t new_seed = static_cast<uint64_t>(r0) | (static_cast<uint64_t>(r1) << 32);
+      uint64_t new_offset = static_cast<uint64_t>(r2) | (static_cast<uint64_t>(r3) << 32);
```
Comment (eee4017) [inline aten/src/ATen/native/cuda/PhiloxKeySplit.cu]: Is there a specific reason for choosing this hash?

File: aten/src/ATen/native/cuda/PhiloxKeySplit.cu [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    int64_t key_idx = tid / num_chunks;
+    int64_t chunk_idx = tid % num_chunks;
+    int64_t split_start = chunk_idx * splits_per_thread;
+    int64_t split_end = min(split_start + splits_per_thread, num_splits);
+
+    uint64_t seed = input[key_idx * 2];
+    uint64_t offset = input[key_idx * 2 + 1];
+
+    curandStatePhilox4_32_10_t state;
+    curand_init(seed, /*subsequence=*/0, /*offset=*/offset, &state);
```
Comment (eee4017) [inline aten/src/ATen/native/cuda/PhiloxKeySplit.cu]: Why not expose the Philox subsequence dimension? Different subsequences occupy disjoint regions of the counter space by construction.

File: aten/src/ATen/native/cuda/PhiloxKeySplit.cu (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/cuda/PhiloxKeySplit.cu]: It's simply reinterpreting the returned 128-bit randomness sampled at this location as a new (64-bit seed, 64-bit offset). This is a convenient way to jump to a location within the 3D PRNG number space (seed, subsequence=0, offset) that is v. likely to result in an independent PRNG stream to avoid undesirable reuse.

File: aten/src/ATen/native/cuda/PhiloxKeySplit.cu (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/cuda/PhiloxKeySplit.cu]: Restricting Philox subsequence to 0 in these kernels (and further not exposing subsequence) is a design decision that helps satisfy some useful properties:
* Consistency of random number generation across GPU types. The underlying kernels for `torch.randn()` / `torch.rand()` today use thread ID-based subsequence numbers, with the # of threads being SM-dependent. This makes the generated outputs different between e.g. A100 vs. H100.
* Consistency of generated values over time AKA "sequential equivalence". This means making the outputs independent of input size / batching, requiring subsequence numbers that are not based on input size; subsequence=0 is the easiest way to achieve this. This property is useful for Distributed, who would like to guarantee that a sharded parameter will have its shards initialized with the same values as if the parameter were initialized all at once. Ex:
```python
key = torch.random.key(42)
vals = torch.random.uniform(key, 10)
sharded_vals1 = torch.random.uniform(key, 5)
# TODO: introduce a public API for this "offset hacking"
offset_key = torch.tensor([42, 5], device=key.device, dtype=key.dtype)
sharded_vals2 = torch.random.uniform(offset_key, 5)
assert torch.equal(vals, torch.cat([sharded_vals1, sharded_vals2], dim=0))
```

Note that ensuring the above properties **does** come at a performance cost. Using thread-based subsequence numbers allows for better memory coalescing behavior and fewer expensive `curand_init()` calls. I think it may be useful to satisfy these properties by default, but provide a flag for opting out for those who care about absolute performance (for this case, simply call into the existing kernels).

File: torch/func/_random.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    Equivalent to ``split(key, data + 1)[data]``, but more efficient when
+    only a single derived key is needed. Useful for associating a key with
+    a loop iteration, layer index, or other integer identifier.
+
+    Supports batched keys: if ``key`` has shape ``(*batch, 2)``, each key in
+    the batch is folded independently.
+
+    Args:
+        key (Tensor): A PRNG key of shape ``(..., 2)`` with dtype ``torch.uint64``.
+        data (int): A non-negative integer to fold into the key.
```
Comment (albanD) [inline torch/func/_random.py]: You don't have any error checking to ensure that it is non-negative?

File: torch/_meta_registrations.py [hunk 4]
```diff
@@ -491,6 +491,17 @@ def meta_rand_default(size, *, dtype=None, layout=None, device=None, pin_memory=
     )
 
 
+@register_meta(aten._philox_key_split.default)
```
Comment (albanD) [inline torch/_meta_registrations.py:494]: Both functions here need to have the same error checking as the kernel with torch._check

File: test/test_stateless_rng.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
+# Owner(s): ["module: random"]
```
Comment (albanD) [inline test/test_stateless_rng.py:1]: Can we add a set of test for oveflows? behaviors?

File: test/test_stateless_rng.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        key1 = random.key(42, device=device)
+        key2 = random.key(42, device=device)
+        self.assertEqual(key1, key2)
+
+    def test_error_unsupported_impl(self, device):
+        with self.assertRaises(NotImplementedError):
+            random.key(42, impl="unsupported", device=device)
+
+
+instantiate_device_type_tests(TestStatelessRNGKey, globals(), only_for=("cuda"))
```
Comment (albanD) [inline test/test_stateless_rng.py]: nit: I think we usually instantiate all of these at the bottom of the file

File: test/test_stateless_rng.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        self.assertEqual(batched.shape, (num_splits, 3, 4, 2))
+        for i in range(3):
+            for j in range(4):
+                individual = random.split(keys[i][j], num_splits)
+                for s in range(num_splits):
+                    self.assertEqual(batched[s][i][j], individual[s])
+
+    def test_error_wrong_shape(self, device):
+        key = torch.tensor([42, 0, 1], dtype=torch.uint64, device=device)
+        with self.assertRaises(RuntimeError):
```
Comment (albanD) [inline test/test_stateless_rng.py]: Can you get claude to use assertRaisesRegex and have the error message here?

File: test/test_stateless_rng.py (hunk 6 above)
Comment (albanD) [inline test/test_stateless_rng.py]:
```suggestion
instantiate_device_type_tests(TestStatelessRNGKey, globals(), only_for=("cuda",))
```
Here and below

File: torch/func/_random.py (hunk 3 above)
Comment (jbschlosser) [reply torch/func/_random.py]: true, it just interprets it as uint64 underneath. updated the docs

File: test/test_stateless_rng.py (hunk 5 above)
Comment (jbschlosser) [reply test/test_stateless_rng.py:1]: sure thing! I have a set of overflow tests for the generation kernels but it's possible that split() and fold_in() can overflow the offset from sampling randomness near the offset UINT64_MAX boundary.

**Review by albanD — APPROVED** [review]
SGTM!

File: aten/src/ATen/native/cuda/PhiloxKeySplit.cu [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  output_sizes.insert(output_sizes.begin(), num_splits);
+  Tensor output = at::empty(output_sizes, key.options());
+  int64_t num_keys = key.numel() / 2;
+  if (num_keys == 0) {
+    return output;
+  }
+
+  int64_t total_threads = num_keys * num_splits;
+  constexpr int block_size = 256;
+  int num_blocks = std::min(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PhiloxKeySplit.cu:95]: btw you don't need to bother with sizing the number of blocks to a particular GPU. Previously because curand_init call was expensive (don't exactly know why and it's not important) we tried to reuse it across grid_stride loop, but all you are doing is `philox_4x32` so there's no reason for a grid-stride loop, just launch enough blocks. IT's a nit for this PR but could simplify the next one.

Comment (jbschlosser) [reply aten/src/ATen/native/cuda/PhiloxKeySplit.cu:95]: nice, love this!
